pkg/crc/network/proxy.go

```diff
@@ -0,0 +1,201 @@
+func (p *ProxyConfig) ProxyConfig() []string {
+	config := []string{}
+
+	if !p.IsEnabled() {
+		return config
+	}
+
+	if p.httpProxy != "" {
+		config = append(config, fmt.Sprintf("HTTP_PROXY=%s", p.httpProxy))
+		config = append(config, fmt.Sprintf("http_proxy=%s", p.httpProxy))
+	}
+
+	if p.httpsProxy != "" {
+		config = append(config, fmt.Sprintf("HTTPS_PROXY=%s", p.httpsProxy))
+		config = append(config, fmt.Sprintf("https_proxy=%s", p.httpsProxy))
+	}
+	config = append(config, fmt.Sprintf("NO_PROXY=%s", p.NoProxy()))
+	config = append(config, fmt.Sprintf("no_proxy=%s", p.NoProxy()))
+
+	return config
+}
+
+// Override proxy for local/intermediate proxy
+func (p *ProxyConfig) OverrideHttpProxy(proxy string) {
+	p.httpProxy = proxy
+}
+
+// Override proxy for local/intermediate proxy
+func (p *ProxyConfig) OverrideHttpsProxy(proxy string) {
+	p.httpsProxy = proxy
+}
+
+// HttpProxy returns the configured value for the HTTP proxy. The empty string is returned in case HTTP proxy is not set.
+func (p *ProxyConfig) HttpProxy() string {
+	return p.httpProxy
+}
+
+// HttpsProxy returns the configured value for the HTTPS proxy. The empty string is returned in case HTTPS proxy is not set.
+func (p *ProxyConfig) HttpsProxy() string {
+	return p.httpsProxy
+}
+
+// NoProxy returns a comma separated list of hosts for which proxies should not be applied.
+func (p *ProxyConfig) NoProxy() string {
+	if p.IsEnabled() {
+		return strings.Join(p.noProxy, ",")
+	} else {
+		return ""
+	}
+}
+
+// AddNoProxy appends the specified host to the list of no proxied hosts.
+func (p *ProxyConfig) AddNoProxy(host string) {
+	p.noProxy = append(p.noProxy, host)
+}
+
+// Sets the current config as environment variables in the current process.
+func (p *ProxyConfig) ApplyToEnvironment() {
+	if !p.IsEnabled() {
+		return
+	}
+
+	if p.httpProxy != "" {
+		os.Setenv("HTTP_PROXY", p.httpProxy)
+		os.Setenv("http_proxy", p.httpProxy)
+	}
+	if p.httpsProxy != "" {
+		os.Setenv("HTTPS_PROXY", p.httpsProxy)
+		os.Setenv("https_proxy", p.httpsProxy)
+	}
+	if len(p.noProxy) != 0 {
+		os.Setenv("NO_PROXY", p.NoProxy())
+		os.Setenv("no_proxy", p.NoProxy())
+	}
+}
+
+// Enabled returns true if at least one proxy (HTTP or HTTPS) is configured. Returns false otherwise.
+func (p *ProxyConfig) IsEnabled() bool {
+	return p.httpProxy != "" || p.httpsProxy != ""
+}
+
+// ValidateProxyURL validates that the specified proxyURL is valid
+func ValidateProxyURL(proxyUrl string, defaultScheme string) error {
+	if proxyUrl == "" {
+		return nil
+	}
+
+	if !strings.HasPrefix(proxyUrl, "http://") &&
+		!strings.HasPrefix(proxyUrl, "https://") {
+		proxyUrl = fmt.Sprintf("%s://%s", defaultScheme, proxyUrl)
+	}
+
+	if !govalidator.IsURL(proxyUrl) {
+		return fmt.Errorf("Proxy URL '%s' is not valid.", proxyUrl)
+	}
+	return nil
+}
+
+// parseProxySpecialChar parse the URI and convert special char to hex
+func parseProxySpecialChar(proxyUrl string, defaultScheme string) string {
```

Comment:
I don't understand why we need to do this :(
`parse` is misleading, `escape` is more descriptive.